### PR TITLE
Preview search with events.

### PR DESCRIPTION
**Pull request by sherakama**
# NOT READY FOR REVIEW
- (Edit the above to reflect status)

# Summary
- TL;DR - what's this PR for?

# Review By (Date)
- When does this need to be reviewed by?

# Criticality
- How critical is this PR on a 1-10 scale? Also see [Severity Assessment](https://stanfordits.atlassian.net/browse/D8CORE-1705).
- E.g., it affects one site, or every site and product?

# Review Tasks

## Setup tasks and/or behavior to test

1. Check out this branch
2. Navigate to...
3. Verify...

## Front End Validation
- [ ] Is the markup using the appropriate semantic tags and passes HTML validation?
- [ ] Cross-browser testing has been performed?
- [ ] Automated accessibility scans performed?
- [ ] Manual accessibility tests performed?
- [ ] Design is approved by @ user?

### Code
- [ ] Are the naming conventions following our standards?
- [ ] Does the code have sufficient inline comments?
- [ ] Is there anything in this code that would be hidden or hard to discover through the UI?
- [ ] Are there any [code smells](https://blog.codinghorror.com/code-smells/)?
- [ ] Are tests provided? eg (unit, behat, or codeception)

### Code security
- [ ] Are all [forms properly sanitized](https://www.drupal.org/docs/8/security/drupal-8-sanitizing-output)?
- [ ] Any obvious [security flaws or new areas for attack](https://www.drupal.org/docs/8/security)?

## General
- [ ] Is there anything included in this PR that is not related to the problem it is trying to solve?
- [ ] Is the approach to the problem appropriate?

# Affected Projects or Products
- Does this PR impact any particular projects, products, or modules?

# Associated Issues and/or People
- JIRA ticket(s)
- Other PRs
- Any other contextual information that might be helpful (e.g., description of a bug that this PR fixes, new functionality that it adds, etc.)
- Anyone who should be notified? (`@mention` them here)

# Resources
- [AMP Tool](https://stanford.levelaccess.net/index.php)
- [Accessibility Manual Test Script](https://docs.google.com/document/d/1ZXJ9RIUNXsS674ow9j3qJ2g1OAkCjmqMXl0Gs8XHEPQ/edit?usp=sharing)
- [HTML Validator](https://validator.w3.org/)
- [Browserstack](https://live.browserstack.com/dashboard) and link to [Browserstack Credentials](https://asconfluence.stanford.edu/confluence/display/SWS/External+Account+Credentials)

### CODE REVIEW

File: gatsby-browser.js [hunk 1]
```diff
@@ -9,6 +9,7 @@ import React from 'react';
 // Contexts.
 import { GlobalStateProvider } from './src/contexts/GlobalContext';
 import { AuthContextProvider } from './src/contexts/AuthContext';
+import { SearchModalProvider } from './src/components/search/Modal/SearchModalContext';
```
Comment (sherakama) [inline gatsby-browser.js:12]: Created a context for the search modal as this made life easier than passing around a bunch of refs.

File: src/utilities/synchronizedEvents.js [hunk 2]
```diff
@@ -1,6 +1,6 @@
 import { DateTime } from 'luxon';
 import TurndownService from 'turndown';
-import markdownConverter from 'storyblok-markdown-richtext';
+import { markdownToRichtext } from 'storyblok-markdown-richtext';
```
Comment (sherakama) [inline src/utilities/synchronizedEvents.js:3]: Fixes a warning.

Comment (ericbakenhus) [reply src/utilities/synchronizedEvents.js:3]: 😬 Their library has export issues and a warning ended up being the least bad option. See my previous attempts:

- #910 
- #911

Comment (sherakama) [reply src/utilities/synchronizedEvents.js:3]: Ah well, that's that then.

File: src/components/search/SearchSuggestions.jsx
Comment (sherakama) [inline src/components/search/SearchSuggestions.jsx:1]: The naming of things in this repo is wild. I'm going to find some time in the future to do some cleanup.

Comment (sherakama) [reply src/components/search/SearchSuggestions.jsx:1]: For now, I am going to change a small amount of stuff and make it worse before it gets better.

File: src/components/content-types/perk/perkPageView.js [hunk 3]
```diff
@@ -2,7 +2,7 @@ import SbEditable from 'storyblok-react';
 import React from 'react';
 import { Heading } from '../../simple/Heading';
 import CardImage from '../../media/cardImage';
-import Layout from '../../partials/layout';
+import Layout from '../../partials/Layout';
```
Comment (sherakama) [inline src/components/content-types/perk/perkPageView.js:5]: The lowercase was really bugging me on this one in particular.

File: src/components/page-types/searchPage.jsx [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+  // eslint-disable-next-line react/destructuring-assignment
+  const itemsPerPageInt = parseInt(props.blok.itemsPerPage, 10);
+
+  // Algolia.
+  // --------------------------------------------------
+  const algoliaClient = algoliasearch(
+    process.env.GATSBY_ALGOLIA_APP_ID,
+    process.env.GATSBY_ALGOLIA_API_KEY
+  );
+  const indexName = 'federated-search-with-events'; // TODO: Change this back before merging.
```
Comment (sherakama) [inline src/components/page-types/searchPage.jsx:24]: Change back to the production index when going live.

File: src/components/page-types/searchPage.jsx
Comment (sherakama) [inline src/components/page-types/searchPage.jsx:1]: Move into using the `InstantSearch` library instead of just the algolia js library.

File: src/utilities/decodeHtmlEntities.js
Comment (sherakama) [inline src/utilities/decodeHtmlEntities.js:1]: New helper utility. Let me know if you think there is a better way.

File: src/components/search/SearchResults.jsx [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+              <SearchResultAlumniEvent result={result} key={result.objectID} />
+            );
+          }
+          default: {
+            return (
+              <SearchResultDefault result={result} key={result.objectID} />
+            );
+          }
+        }
+      })}
```
Comment (sherakama) [inline src/components/search/SearchResults.jsx:63]: I split up the templates for the different types as this makes it much easier to extend/theme.

File: src/components/search/SearchResults.jsx [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+        <Skeleton variant="rectangular" className="su-rs-my-1" height={135} />
+        <Skeleton variant="rectangular" className="su-rs-my-1" height={135} />
+        <Skeleton variant="rectangular" className="su-rs-my-1" height={135} />
+        <Skeleton variant="rectangular" className="su-rs-my-1" height={135} />
+        <Skeleton variant="rectangular" className="su-rs-my-1" height={135} />
+        <Skeleton variant="rectangular" className="su-rs-my-1" height={135} />
+        <Skeleton variant="rectangular" className="su-rs-my-1" height={135} />
+      </div>
+    );
+  }
```
Comment (sherakama) [inline src/components/search/SearchResults.jsx:31]: A loading state was asked for as the 'thinking' wasn't apparent before.

File: src/components/search/SearchResults.jsx [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+
+  // No Results.
+  if (!results || !results.nbHits) {
+    return null;
+  }
```
Comment (sherakama) [inline src/components/search/SearchResults.jsx:36]: There is a no results component on the SearchPage component for this condition.

File: src/components/partials/Layout.jsx
Comment (sherakama) [inline src/components/partials/Layout.jsx:1]: At some point I will change all of our files. For now, I'll just add to the mess and change this one.

File: src/components/search/Hits/SearchResultAlumniEvent.jsx
Comment (sherakama) [inline src/components/search/Hits/SearchResultAlumniEvent.jsx:1]: New template for the search result type of `event`

File: src/components/search/Hits/SearchResultAlumniEvent.jsx [hunk 8]
```diff
@@ -0,0 +1,128 @@
...
+        return `${startTime.toFormat('EEEE, LLLL d, yyyy h:mm a ZZZZ')}`;
+      }
+      return `${startTime.toFormat(
+        'EEEE, LLLL d, yyyy h:mm a ZZZZ'
+      )} - ${endTime.toFormat('h:mm a ZZZZ')}`;
+    }
+    return `${startTime.toFormat('LLLL d, yyyy h:mm a')} - ${endTime.toFormat(
+      'LLLL d, yyyy h:mm a ZZZZ'
+    )}`;
+  };
```
Comment (sherakama) [inline src/components/search/Hits/SearchResultAlumniEvent.jsx:64]: @ericbakenhus, you've done quite a bit with the formatting and handling of these dates in the event search. I've got client approval on how it looks but is there anything I am missing/should know?

File: src/components/search/Mobile/MobileFilterFooter.jsx
Comment (sherakama) [inline src/components/search/Mobile/MobileFilterFooter.jsx:1]: This is mostly a copy of the one from the events-search but with different clearRefinements handling.

File: src/components/search/Mobile/MobileFilterHeader.jsx
Comment (sherakama) [inline src/components/search/Mobile/MobileFilterHeader.jsx:1]: This is a copy of the one from events-search. I didn't feel they were tied together enough to warrant using the component directly from the events-search but I could be convinced otherwise.

File: src/components/search/Modal/SearchFieldModal.jsx [hunk 9]
```diff
@@ -0,0 +1,206 @@
...
+ * @type {React.FC<Props>}
+ * @returns {React.ReactElement}
+ */
+const SearchFieldModal = React.forwardRef(({ emptySearchMessage }, ref) => {
+  // Algolia Client.
+  const algoliaClient = algoliasearch(
+    process.env.GATSBY_ALGOLIA_APP_ID,
+    process.env.GATSBY_ALGOLIA_API_KEY
+  );
+  const indexName = 'federated-search-with-events'; // TODO: CHANGE THIS BACK BEFORE MERGING
```
Comment (sherakama) [inline src/components/search/Modal/SearchFieldModal.jsx:21]: Change this back.

File: src/components/search/Modal/SearchFieldModal.jsx [hunk 10]
```diff
@@ -0,0 +1,206 @@
...
+/**
+ * @type {React.FC<Props>}
+ * @returns {React.ReactElement}
+ */
+const SearchFieldModal = React.forwardRef(({ emptySearchMessage }, ref) => {
+  // Algolia Client.
+  const algoliaClient = algoliasearch(
+    process.env.GATSBY_ALGOLIA_APP_ID,
+    process.env.GATSBY_ALGOLIA_API_KEY
+  );
```
Comment (sherakama) [inline src/components/search/Modal/SearchFieldModal.jsx:20]: Suppose I could have created a getClient utility...

File: src/components/search/Modal/SearchFieldModal.jsx
Comment (sherakama) [inline src/components/search/Modal/SearchFieldModal.jsx:1]: This changes the modal search field to look and behave more like the other search fields we have around this site. The search suggestions have been replaced with an autocomplete option.

File: src/components/search/Hits/SearchResultAlumniEvent.jsx (hunk 8 above)
Comment (ericbakenhus) [reply src/components/search/Hits/SearchResultAlumniEvent.jsx:64]: Are the `start` and `end` strings directly from Algolia/Storyblok? If so, they're not in ISO format (thanks Storyblok). You'd want to use our `luxonDate` utility function instead of `Datetime.fromISO`. You'll run into weird Invalid Date issues otherwise (e.g. in "old" Safari versions).

File: src/components/search/SearchField.jsx [hunk 11]
```diff
@@ -0,0 +1,209 @@
...
+ * @type {React.FC<Props>}
+ * @returns {React.ReactElement}
+ */
+const SearchField = ({ emptySearchMessage }) => {
+  // Algolia Client.
+  const algoliaClient = algoliasearch(
+    process.env.GATSBY_ALGOLIA_APP_ID,
+    process.env.GATSBY_ALGOLIA_API_KEY
+  );
+  const indexName = 'federated-search-with-events'; // TODO: CHANGE THIS BACK BEFORE MERGING
```
Comment (sherakama) [inline src/components/search/SearchField.jsx:19]: Change this here too.

File: src/components/search/Hits/SearchResultAlumniEvent.jsx (hunk 8 above)
Comment (sherakama) [reply src/components/search/Hits/SearchResultAlumniEvent.jsx:64]: > Re: ISO Format

I convert them to ISO before shipping them to Algolia in the middleware.

https://github.com/SU-SWS/adapt-algolia-index/pull/120/files#diff-96e41588c18a3e48cc1df57393f8a9ce693cf4cee398198c5c83b27243a3cbf3R75-R77

> You'd want to use our luxonDate utility function

Ah right. Thanks!

**Comment by sherakama** [comment]
I'll look into the failing tests tomorrow.

**Comment by ericbakenhus** [comment]
@sherakama Part 1 reviewed. Buttons are buttoning, modals are modaling, searches are searching, focus traps are trapping focus. 👍 

One (very) subjective suggestion: when it auto scrolls to the search input on the search page it feels a little cramped at the top of the window. We could give it a little breathing room by tossing a top scroll margin on it.

No scroll margin:
Screenshot 2024-10-24 at 5 30 58 PM

With 50px top scroll margin:
Screenshot 2024-10-24 at 5 30 48 PM

**Comment by ericbakenhus** [comment]
@sherakama Part 2: mobile seems good. 👍

**Comment by ericbakenhus** [comment]
@sherakama Part 3. Ran into a few issues. Everything not mentioned tested exactly as expected though. 👍 

Issues found:

- Seems like using the filter skip nav (i.e. adding a hash to the URL) triggers a new search/loading state.
- If the URL has a hash in it (e.g. after using the skip filter nav), the clear search button doesn't reset the URL query or hash.
- Bit an a11y/UX issue, selecting a filter makes the focus jump over to the wrapper element.
- I'm not scrolled to the top after pagination (in FF). For the events search, I had to toss the scrolling code in a `setTimeout` to get it correctly queued up after the page change for events.
- If I have an existing search (e.g. URL has `q=test` and results are showing for that search), but then manually clear the search input and hit enter/activate the search button then: 
    - the "Uh oh..." message _does_ appear, but 
    - the"clear search" button also disappears, which makes clearing the existing search difficult.
- I see an expired event in the results:

Screenshot 2024-10-25 at 11 58 01 AM

We're adding a date-based filter onto the events search to prevent that (I don't know if that can be easily applied here; I haven't looked at code yet). It should be less of an issue in production since the daily import function _should_ be unpublishing/moving expired events.

- I see an empty event. Don't know what's going on with that. 😅 

Screenshot 2024-10-25 at 11 57 40 AM


**_Subjective_** thoughts:

- Like before, maybe we could add some top scroll margin to the top of the results/wrapper/whatever it moves focus to (for when following the skip filters nav)
- I _really_ dislike using loading skeletons to show when Algolia is fetching an uncached query (searching on Directory gives me an aneurysm). The events search, for example, doesn't show any loading indicator (except if it hangs, then we toss on a 50% opacity and show a loading spinner). Even if it was slower (looking at you Coveo), I'd still heavily favor the opacity and/or spinner approach.
- That wine banner is **huge**. More of a wine header. 😆

**Comment by sherakama** [comment]
> * Seems like using the filter skip nav (i.e. adding a hash to the URL) triggers a new search/loading state.

Huh yeah, that's strange. I'll look into that.

> * If the URL has a hash in it (e.g. after using the skip filter nav), the clear search button doesn't reset the URL query or hash.

Thanks, I'll look into that.

> * Bit an a11y/UX issue, selecting a filter makes the focus jump over to the wrapper element.

I'll keep it on the filter. 

> * I'm not scrolled to the top after pagination (in FF). For the events search, I had to toss the scrolling code in a `setTimeout` to get it correctly queued up after the page change for events.

Thanks, I'll see what I can do.

> * If I have an existing search (e.g. URL has `q=test` and results are showing for that search), but then manually clear the search input and hit enter/activate the search button then:
>   * the "Uh oh..." message _does_ appear, but
>   * the"clear search" button also disappears, which makes clearing the existing search difficult.

Good catch

> We're adding a date-based filter onto the events search to prevent (expired events)

That would be a good belt and suspenders thing to add.

> * I see an empty event. Don't know what's going on with that. 😅
 
I caught that too. Added some more validation in the middleware: https://github.com/SU-SWS/adapt-algolia-index/pull/120/commits/6fb73747a0e913f32ae4a14fbecfe22740eb1f3c 

> * Like before, maybe we could add some top scroll margin to the top of the results/wrapper/whatever it moves focus to (for when following the skip filters nav)

The scroll kinda bugs me overall. Perhaps I just remove it. 

> * I _really_ dislike using loading skeletons to show when Algolia is fetching an uncached query (searching on Directory gives me an aneurysm). The events search, for example, doesn't show any loading indicator (except if it hangs, then we toss on a 50% opacity and show a loading spinner). Even if it was slower (looking at you Coveo), I'd still heavily favor the opacity and/or spinner approach.

Yeah, maybe I'll two step this. Opacity immediately, and skeles for hung. 

> * That wine banner is **huge**. More of a wine hero image/header. 😆

Yuuuuuge

File: src/components/events-discovery/eventsDiscovery.jsx [hunk 12]
```diff
@@ -120,7 +120,7 @@ const EventDiscoveryContent = () => {
               <Hits
                 hitComponent={Hit}
                 classNames={{
-                  root: `${isStalled ? 'su-opacity-50' : ''}`,
+                  root: `${isStalled ? 'su-opacity-30' : ''}`,
```
Comment (sherakama) [inline src/components/events-discovery/eventsDiscovery.jsx:123]: @ericbakenhus 50 didn't seem like enough on my monitor.

File: src/components/search/SearchFacet.jsx [hunk 13]
```diff
@@ -1,23 +1,23 @@
...
+  const isStalled = status === 'stalled';
 
   // Filter out any items we don't want to show.
   const filteredItems = items.filter((item) => !excludes.includes(item.value));
   // Sort the items alphabetically.
   filteredItems.sort((a, b) => a.value.localeCompare(b.value));
 
   // Show loading.
-  if (status === 'loading') {
+  if (isStalled) {
```
Comment (sherakama) [inline src/components/search/SearchFacet.jsx:20]: Set the skeles to show when stalled (>2s)

Before that, the opacity is set to 30.

File: src/components/page-types/searchPage.jsx [hunk 14]
```diff
@@ -56,11 +56,12 @@ const SearchPage = (props) => {
     <InstantSearch
       searchClient={algoliaClient}
       indexName={indexName}
-      stalledSearchDelay={3000}
+      stalledSearchDelay={2000}
```
Comment (sherakama) [inline src/components/page-types/searchPage.jsx:61]: Stall and show the skeles at 2s

File: src/components/search/SearchField.jsx [hunk 15]
```diff
@@ -20,13 +20,14 @@ const SearchField = ({ emptySearchMessage }) => {
   const index = algoliaClient.initIndex(indexName);
 
   // Hooks and state.
-  const { query, refine } = useSearchBox();
+  const { query, refine, clear } = useSearchBox();
```
Comment (sherakama) [inline src/components/search/SearchField.jsx:23]: Hey, would you look at that! A clear function. Might be a good fit for the clear link? 🤦

File: src/components/search/SearchField.jsx [hunk 16]
```diff
@@ -44,14 +45,27 @@ const SearchField = ({ emptySearchMessage }) => {
         return;
       }
 
+      // Add the filters to the autocomplete as well.
```
Comment (sherakama) [inline src/components/search/SearchField.jsx:41]: I've patched in the facets so that they filter the autocomplete results and don't show anything outside of what is available. Before this, if I filtered down and then used the autocomplete to select an item outside of the current filter range, I got a no results found message.

File: src/components/accessibility/Skiplink.jsx
Comment (sherakama) [inline src/components/accessibility/Skiplink.jsx:1]: I'm not sure if I've committed a crime here or not but this is helping the search page as the anchor links are triggering gatsby's (reach) router to do a 'navigate' and render. The change to the URL is also picked up by the `history` function in the instantsearch router.

By putting the path into the URL and omitting the default browser behaviour I skip the route page render but I am still getting the `history` event and the search still searches. At least with this code the search and facets still update the URL when they are interacted with.

Comment (ericbakenhus) [reply src/components/accessibility/Skiplink.jsx:1]: It results in all the same expected behavior, so I'd say it's legal. 😆

File: src/components/accessibility/Skiplink.jsx [hunk 17]
```diff
@@ -16,9 +18,22 @@ export const Skiplink = ({
...
   return (
     <a
       href={anchorLink}
+      onClick={(e) => {
+        e.preventDefault();
+        navigate(currentUrl + anchorLink);
+        const elem = document.getElementById(anchorLink.replace('#', ''));
+        elem.setAttribute('tabindex', -1);
+        elem.focus({ preventScroll: true });
+        scrollTo(anchorLink);
```
Comment (ericbakenhus) [inline src/components/accessibility/Skiplink.jsx:35]: The -1 tab index will also remove elements form the tab order if they're focuable by default. I don't think we're doing that anywhere currently, but that could pitfall for a future developer.

We could always store and restore the existing tab index:

```suggestion
        const prevTabIndex = elem.tabIndex;
        elem.setAttribute('tabindex', -1);
        elem.focus({ preventScroll: true });
        elem.setAttribute('tabindex', prevTabIndex);
        scrollTo(anchorLink);
```

Comment (sherakama) [reply src/components/accessibility/Skiplink.jsx]: We weren't previously skipping to anything that was tabbable but good call out on the future dev pitfall. Thanks for the GH suggestion.

**Comment by sherakama** [comment]
@ericbakenhus and @yvonnetangsu I feel 'good enough' with this version. I've added a couple of follow-up tickets for the next sprint. Thanks for all of your review and feedback

[DS-1027](https://stanford.atlassian.net/browse/DS-1027)
[DS-1028](https://stanford.atlassian.net/browse/DS-1028)
[DS-1029](https://stanford.atlassian.net/browse/DS-1029)

**Comment by sherakama** [comment]
Ah shoot. Forgot to change the index back!
